opencog/persist/zmq/atomspace/ZMQPersistSCM.cc

```diff
@@ -0,0 +1,227 @@
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <opencog/atomspace/AtomSpace.h>
+#include <opencog/atomspace/BackingStore.h>
+#include <opencog/guile/SchemePrimitive.h>
+#include <opencog/persist/zmq/atomspace/ZMQClient.h>
+
+#include "ZMQPersistSCM.h"
+
+using namespace opencog;
+
+namespace opencog
+{
+/** \addtogroup grp_persist
+ *  @{
+ */
+
+class ZMQBackingStore : public BackingStore
+{
+	private:
+		ZMQClient *_store;
+	public:
+		ZMQBackingStore();
+		void set_store(ZMQClient *);
+
+		virtual NodePtr getNode(Type, const char *) const;
+		virtual LinkPtr getLink(Type, const HandleSeq&) const;
+		virtual AtomPtr getAtom(Handle) const;
+		virtual HandleSeq getIncomingSet(Handle) const;
+		virtual void storeAtom(Handle);
+		virtual void loadType(AtomTable&, Type);
+		virtual void barrier();
+};
+
+/** @}*/
+}  // namespace
+
+ZMQBackingStore::ZMQBackingStore()
+{
+	_store = NULL;
+}
+
+void ZMQBackingStore::set_store(ZMQClient *as)
+{
+	_store = as;
+}
+
+NodePtr ZMQBackingStore::getNode(Type t, const char *name) const
+{
+	return _store->getNode(t, name);
+}
+
+LinkPtr ZMQBackingStore::getLink(Type t, const std::vector<Handle>& oset) const
+{
+	return _store->getLink(t, oset);
+}
+
+AtomPtr ZMQBackingStore::getAtom(Handle h) const
+{
+	return _store->getAtom(h);
+}
+
+HandleSeq ZMQBackingStore::getIncomingSet(Handle h) const
+{
+	return _store->getIncomingSet(h);
+}
+
+void ZMQBackingStore::storeAtom(Handle h)
+{
+	_store->storeAtom(h);
+}
+
+void ZMQBackingStore::loadType(AtomTable& at, Type t)
+{
+	_store->loadType(at, t);
+}
+
+void ZMQBackingStore::barrier()
+{
+	_store->flushStoreQueue();
+}
+
+// ==================================================================
+
+ZMQPersistSCM::ZMQPersistSCM(AtomSpace *as)
+{
+	_as = as;
+	_store = NULL;
+	_backing = new ZMQBackingStore();
+
+#ifdef HAVE_GUILE
+	static bool is_init = false;
+	if (is_init) return;
+	is_init = true;
+	scm_with_guile(init_in_guile, this);
+#endif
+}
+
+void* ZMQPersistSCM::init_in_guile(void* self)
+{
+#ifdef HAVE_GUILE
+	scm_c_define_module("opencog persist-zmq", init_in_module, self);
+	scm_c_use_module("opencog persist-zmq");
+#endif
+	return NULL;
+}
+
+void ZMQPersistSCM::init_in_module(void* data)
+{
+   ZMQPersistSCM* self = (ZMQPersistSCM*) data;
+   self->init();
+}
+
+void ZMQPersistSCM::init(void)
+{
+#ifdef HAVE_GUILE
+	define_scheme_primitive("zmq-open", &ZMQPersistSCM::do_open, this, "persist-zmq");
+	define_scheme_primitive("zmq-close", &ZMQPersistSCM::do_close, this, "persist-zmq");
+	define_scheme_primitive("zmq-load", &ZMQPersistSCM::do_load, this, "persist-zmq");
+	define_scheme_primitive("zmq-store", &ZMQPersistSCM::do_store, this, "persist-zmq");
+#endif
+}
+
+ZMQPersistSCM::~ZMQPersistSCM()
+{
+	delete _backing;
+}
+
+void ZMQPersistSCM::do_open(const std::string& networkAddress)
+{
+	_store = new ZMQClient(networkAddress);
```

Comment:
I'm not talking about zmq, I'm talking about Neo4J.  When you go to access the database itself, via zmq or via rest or via whatever, you have to be able to prove that you are you, as otherwise,  any russian botnet could log into your database and rape it.  There has to be some way of protecting your data from hostile actors....